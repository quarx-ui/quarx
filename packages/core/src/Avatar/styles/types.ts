import { PickSxBorderSize, SxSize } from '@core/enums';

export type AvatarSize = Exclude<SxSize, 'base'>
export type AvatarBorderRadius = PickSxBorderSize<'rounded' | 'smooth' | 'square'>;

export interface AvatarStyleParams {
    size: AvatarSize,
    radius: AvatarBorderRadius,
    showBorder: boolean,
}
