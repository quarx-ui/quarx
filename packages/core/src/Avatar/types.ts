import { ComponentPropsWithHTML, WithClassesAndStyles } from '@core';
import { ReactNode } from 'react';
import { AvatarStyleKeys, AvatarStyleParams } from './styles';

export interface AvatarPropsWithoutHtml extends
    Partial<AvatarStyleParams>,
    WithClassesAndStyles<AvatarStyleKeys, AvatarStyleParams>,
    AvatarImageProps
{
    className?: string,
    children?: ReactNode,
}

export type AvatarImagePropKeys =
| 'alt'
| 'src'
| 'srcSet'
| 'crossOrigin'
| 'referrerPolicy'

export type AvatarImageProps = Pick<JSX.IntrinsicElements['img'], AvatarImagePropKeys>;
export type AvatarProps = ComponentPropsWithHTML<AvatarPropsWithoutHtml>;
export * from './styles/types';
