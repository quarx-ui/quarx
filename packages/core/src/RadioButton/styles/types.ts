import { SelectionSize } from '@core/src/Selection/types';

export interface RadioButtonStyleParams {
    /** Состояние наведения */
    hover?: boolean,

    /** Состояние компонента */
    checked?: boolean,

    /** Наличие ошибки */
    hasError?: boolean,

    /** Размер компонента */
    size?: SelectionSize,

    /** Отключение возможности фокуса */
    disableFocus?: boolean,

    /** Изменяет состояние компонента на активное/неактивное */
    disabled?: boolean,
}
