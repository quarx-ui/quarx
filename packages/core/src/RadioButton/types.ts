import { ChangeEventHandler, ReactNode, RefObject } from 'react';
import { WithClassesAndStyles } from '@core';
import { BaseProps } from '@core/types';
import { SelectionProps } from '@core/src';
import { SelectionSize } from '@core/src/Selection/types';
import { RadioButtonStyleKeys } from './style';

export interface RadioButtonStyleParams {
    /** Состояние наведения */
    hover?: boolean,

    /** Состояние компонента */
    checked?: boolean,

    /** Наличие ошибки */
    hasError?: boolean,

    /** Размер компонента */
    size?: SelectionSize,

    /** Отключение возможности фокуса */
    disableFocus?: boolean,

    /** Изменяет состояние компонента на активное/неактивное */
    disabled?: boolean,
}

export interface RadioButtonPropsWithoutHtml extends
    RadioButtonStyleParams,
    WithClassesAndStyles<RadioButtonStyleKeys, RadioButtonStyleParams>,
    BaseProps<HTMLLabelElement>
{
    /** Дочерний элемент */
    children?: ReactNode,

    /** Объект со свойствами для элемента `input` */
    inputProps?: JSX.IntrinsicElements['input'],

    /** Ссылка к элементу `input` */
    inputRef?: RefObject<HTMLInputElement>,

    /** Обработчик изменения состояния элемента `input` */
    onChange?: ChangeEventHandler<HTMLInputElement>,

    /** Свойство `name` элемента `input` */
    name?: string,

    /** Свойство `value` элемента `input` */
    value?: string,
}

export type RadioButtonHtmlAttributes = Omit<JSX.IntrinsicElements['label'], keyof RadioButtonPropsWithoutHtml>
export type RadioButtonProps = RadioButtonPropsWithoutHtml & RadioButtonHtmlAttributes;

export interface RadioButtonSelectionProps extends
    SelectionProps,
    Pick<RadioButtonProps, 'checked'>
{
    /** Объект с параметрами для компонента `checkbox` */
    radioButtonProps?: RadioButtonProps,
}
