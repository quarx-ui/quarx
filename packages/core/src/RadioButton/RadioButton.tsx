/** @jsx jsx */
import { jsx } from '@emotion/react';
import React, { FC, forwardRef, MouseEventHandler, useState } from 'react';
import { SX_SIZE } from '@core/enums';
import { usePropsOverwrites } from '@core';
import { useStyles } from './styles';
import { RadioButtonProps } from './types';

export const RadioButton:FC<RadioButtonProps> = forwardRef<HTMLLabelElement, RadioButtonProps>((
    initialProps,
    ref,
) => {
    const { props, cn } = usePropsOverwrites('RadioButton', initialProps);
    const {
        children,
        hasError = false,
        styles: externalStyles,
        disableFocus = false,
        size = SX_SIZE.medium,
        disabled = false,
        checked,
        hidden = false,
        hover: externalHover,
        onChange,
        inputRef,
        inputProps,
        name,
        value,
        onMouseEnter,
        onMouseLeave,
        ...restProps
    } = props;

    const [hover, setHover] = useState(externalHover ?? false);

    const params = {
        size,
        disableFocus,
        hover: externalHover ?? hover,
        hasError,
        disabled,
        checked: checked ?? false,
    };

    const styles = useStyles({ ...params, styles: externalStyles });

    if (hidden) {
        return null;
    }

    const mouseEnterHandler: MouseEventHandler<HTMLLabelElement> = (e) => {
        setHover(true);
        onMouseEnter?.(e);
    };

    const mouseLeaveHandler: MouseEventHandler<HTMLLabelElement> = (e) => {
        setHover(false);
        onMouseLeave?.(e);
    };

    return (
        <label
            ref={ref}
            className={cn('root', params)}
            css={styles.root}
            onMouseEnter={mouseEnterHandler}
            onMouseLeave={mouseLeaveHandler}
            {...restProps}
        >
            <input
                ref={inputRef}
                type="radio"
                name={name}
                value={value}
                className={cn('input')}
                css={styles.input}
                disabled={disabled}
                checked={checked}
                onChange={onChange}
                {...inputProps}
            />
            <div
                className={cn('marker')}
                css={styles.marker}
            >
                <div
                    className={cn('markerDot')}
                    css={styles.markerDot}
                />
            </div>
            {children && (
                <div
                    className={cn('content')}
                    css={styles.content}
                >
                    {children}
                </div>
            )}
        </label>
    );
});
