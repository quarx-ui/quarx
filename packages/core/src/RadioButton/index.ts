export { RadioButton } from './RadioButton';
export type { RadioButtonProps, RadioButtonSelectionProps } from './types';
export { RadioButtonSelection } from './RadioButtonSelection';
