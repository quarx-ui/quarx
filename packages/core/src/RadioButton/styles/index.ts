import { KeysFromUseStyles, makeStyles, typography } from '@core';
import { paramsToCss } from '@core/utils/paramsToCss';
import { RadioButtonStyleParams } from './types';

export const useStyles = makeStyles((
    { palette, transitions },
    { size, hover, disableFocus, hasError, disabled, checked }: Required<RadioButtonStyleParams>,
) => ({
    root: [
        {
            position: 'relative',
            display: 'flex',
            alignItems: 'center',
            backgroundColor: 'transparent',
            cursor: 'pointer',
        },
        disabled && {
            cursor: 'not-allowed',
        },
    ],

    input: [
        {
            position: 'absolute',
            appearance: 'none',
        },
        !disableFocus && {
            '&:focus-visible + .SxRadioButton-marker': {
                borderWidth: 2,
                borderColor: palette.border.focus.dark,
            },
        },
        !checked && !hasError && !disableFocus && !disabled && {
            '&:focus-visible + .SxRadioButton-marker': {
                backgroundColor: 'transparent',
            },
        },
        checked && !hasError && !disableFocus && {
            '&:focus-visible + .SxRadioButton-marker': {
                backgroundColor: palette.colors.brand.default,
            },
        },
        checked && hasError && !disableFocus && {
            '&:focus-visible + .SxRadioButton-marker': {
                backgroundColor: palette.colors.danger.default,
            },
        },
    ],

    marker: [
        {
            borderRadius: '100%',
            boxSizing: 'border-box',
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            // backgroundColor: palette.RadioButton.enabled.bg,
            border: `1px solid ${palette.border.secondary}`,
            transition: transitions.create(['background-color', 'border-color']),
        },
        paramsToCss(size)({
            small: {
                height: 16,
                width: 16,
            },
            medium: {
                height: 18,
                width: 18,
            },
            large: {
                height: 20,
                width: 20,
            },
        }),

        checked && {
            backgroundColor: palette.colors.brand.default,
            borderColor: 'transparent',
        },
        hover && checked && {
            backgroundColor: palette.colors.brand.hover,
        },
        hasError && {
            borderColor: palette.colors.danger.default,
        },
        hasError && checked && {
            backgroundColor: palette.colors.danger.default,
        },
        (disabled || (disabled && hasError)) && {
            backgroundColor: palette.text.tertiary,
            borderColor: 'transparent',
        },
        disabled && checked && {
            backgroundColor: palette.text.tertiary,
        },
    ],
    markerDot: [
        {
            position: 'absolute',
            borderRadius: '100%',
            width: 8,
            height: 8,
            backgroundColor: 'transparent',
            transition: transitions.create('background-color'),
        },
        size === 'small' && {
            width: 6,
            height: 6,
        },
        disabled && {
            backgroundColor: 'transparent',
        },
        (checked || (checked && !hasError) || (checked && disabled))
            && {
                backgroundColor: palette.text.constant,
            },
        !checked && !disabled && hover && {
            backgroundColor: palette.border.secondary,
        },
    ],
    content: [
        {
            marginLeft: 12,
        },
        paramsToCss(size)({
            small: typography.Text.M.Regular,
            medium: typography.Text.L.Regular,
            large: typography.Text.XL.Regular,
        }),
    ],
}), { name: 'SxRadioButton' });

export type RadioButtonStyleKeys = KeysFromUseStyles<typeof useStyles>

export * from './types';
