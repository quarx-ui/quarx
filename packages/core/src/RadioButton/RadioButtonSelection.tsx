import React, { createRef, FC, forwardRef, MouseEventHandler, useState } from 'react';
import { Selection } from '@core/src';
import { RadioButton, usePropsOverwrites } from '@core';
import { RadioButtonSelectionProps } from '@core/src/RadioButton/types';

export const RadioButtonSelection: FC<RadioButtonSelectionProps> = forwardRef<
HTMLDivElement,
RadioButtonSelectionProps
>((
    initialProps,
    ref,
) => {
    const { props, cn } = usePropsOverwrites('RadioButtonSelection', initialProps);
    const {
        disabled,
        size,
        focused: externalFocus,
        onMouseLeave,
        onMouseEnter,
        radioButtonProps,
        checked,
    } = props;

    const [focus, setFocus] = useState(externalFocus ?? false);
    const [hover, setHover] = useState(false);

    const mouseLeaveHandler: MouseEventHandler<HTMLDivElement> = (e) => {
        setHover(false);
        onMouseLeave?.(e);
    };

    const mouseEnterHandler: MouseEventHandler<HTMLDivElement> = (e) => {
        setHover(true);
        onMouseEnter?.(e);
    };

    const radioButtonRef = createRef<HTMLInputElement>();

    return (
        <Selection
            {...props}
            className={cn('root')}
            onMouseEnter={mouseEnterHandler}
            onMouseLeave={mouseLeaveHandler}
            disabled={radioButtonRef.current?.disabled || disabled}
            focused={externalFocus ?? focus}
            ref={ref}
            leftAdornment={(
                <RadioButton
                    inputRef={radioButtonRef}
                    disabled={radioButtonRef.current?.disabled || disabled}
                    disableFocus
                    color={radioButtonProps?.color}
                    checked={checked}
                    hover={hover}
                    size={size}
                    {...radioButtonProps}
                    onFocus={() => setFocus(true)}
                    onBlur={() => setFocus(false)}
                />
            )}
        />
    );
});
