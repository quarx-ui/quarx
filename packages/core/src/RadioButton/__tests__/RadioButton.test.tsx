import { cleanup, render, screen } from '@testing-library/react';
import { RadioButtonStyleParams } from '@core/src/RadioButton/types';
import React, { FC } from 'react';
import { RadioButton } from '@core';
import userEvent from '@testing-library/user-event';
import { expectPropsMapInClasses } from '../../../../../test-utils/jest';

const checkPropsInClasses = (props: RadioButtonStyleParams) => {
    const {
        size = 'medium',
        hasError = false,
        disableFocus = false,
        disabled = false,
    } = props;
    const radioButton = document.querySelector('label');
    if (!radioButton) { return; }
    const propsWithDefault = { size, disabled, hasError, disableFocus };

    expectPropsMapInClasses(radioButton)(propsWithDefault);
};

const checkProps = (Component: FC) => (checkedProps?: RadioButtonStyleParams) => () => {
    const { asFragment } = render(<Component {...checkedProps}>{Component.displayName}</Component>);

    checkPropsInClasses(checkedProps ?? {});
    expect(asFragment()).toMatchSnapshot();
};

describe('RadioButton snapshots', () => {
    RadioButton.displayName = 'RadioButton';
    const checkRadioProps = checkProps(RadioButton);

    it('default', checkRadioProps());

    it('small', checkRadioProps({ size: 'small' }));
    it('medium', checkRadioProps({ size: 'medium' }));
    it('large', checkRadioProps({ size: 'large' }));

    it('checked', checkRadioProps({ checked: true }));
    it('disabled', checkRadioProps({ disabled: true }));
    it('disableFocus', checkRadioProps({ disableFocus: true }));
    it('hasError', checkRadioProps({ hasError: true }));
});

describe('RadioButton', () => {
    it('text should be in the document', () => {
        render(<RadioButton>RadioButton</RadioButton>);

        expect(screen.queryByText('RadioButton')).toBeInTheDocument();
        const radioButton = document.querySelector('input');
        expect(radioButton).not.toHaveAttribute('disabled');
    });

    it('radioButton should be checked', () => {
        render(<RadioButton checked>RadioButton</RadioButton>);

        const radioButton = document.querySelector('input');
        expect(radioButton?.checked).toEqual(true);
    });

    it('radioButton should have attribute disabled', () => {
        const mockedOnClick = jest.fn();
        render(<RadioButton disabled onClick={mockedOnClick}>RadioButton</RadioButton>);
        const radioButton = document.querySelector('input');

        if (!radioButton) { return; }

        expect(radioButton).toHaveAttribute('disabled');
        userEvent.click(radioButton);
        expect(mockedOnClick).not.toBeCalled();
    });

    it('radioButton should be hidden', () => {
        render(<RadioButton hidden>RadioButton</RadioButton>);
        expect(document.querySelector('input')).not.toBeInTheDocument();

        cleanup();

        render(<RadioButton permissions={{ hidden: true }}>RadioButton</RadioButton>);
        expect(document.querySelector('input')).not.toBeInTheDocument();
    });

    it('onClick should be called', () => {
        let checked = false;
        const mockedOnClick = jest.fn(() => {
            checked = !checked;
        });
        render(<RadioButton onClick={mockedOnClick}>RadioButton</RadioButton>);
        const radioButton = document.querySelector('input');

        if (!radioButton) { return; }

        userEvent.click(radioButton);
        expect(mockedOnClick).toBeCalledTimes(1);
        expect(checked).toEqual(true);
    });
});
