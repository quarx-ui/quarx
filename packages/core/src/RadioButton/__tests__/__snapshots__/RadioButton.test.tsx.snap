// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton snapshots checked 1`] = `
<DocumentFragment>
  <label
    class="SxRadioButton SxRadioButton_size_medium SxRadioButton_checked css-43x8eg-SxRadioButton-root"
  >
    <input
      checked=""
      class="SxRadioButton-input css-1hik84s-SxRadioButton-input"
      type="radio"
      value=""
    />
    <div
      class="SxRadioButton-marker css-onplbh-SxRadioButton-marker"
    >
      <div
        class="SxRadioButton-markerDot css-3mtgan-SxRadioButton-markerDot"
      />
    </div>
    <div
      class="SxRadioButton-content css-1rk6f29-SxRadioButton-content"
    >
      RadioButton
    </div>
  </label>
</DocumentFragment>
`;

exports[`RadioButton snapshots default 1`] = `
<DocumentFragment>
  <label
    class="SxRadioButton SxRadioButton_size_medium css-43x8eg-SxRadioButton-root"
  >
    <input
      class="SxRadioButton-input css-4aocnp-SxRadioButton-input"
      type="radio"
      value=""
    />
    <div
      class="SxRadioButton-marker css-1389et7-SxRadioButton-marker"
    >
      <div
        class="SxRadioButton-markerDot css-nhzvrs-SxRadioButton-markerDot"
      />
    </div>
    <div
      class="SxRadioButton-content css-1rk6f29-SxRadioButton-content"
    >
      RadioButton
    </div>
  </label>
</DocumentFragment>
`;

exports[`RadioButton snapshots disableFocus 1`] = `
<DocumentFragment>
  <label
    class="SxRadioButton SxRadioButton_size_medium SxRadioButton_disableFocus css-43x8eg-SxRadioButton-root"
  >
    <input
      class="SxRadioButton-input css-irpnfb-SxRadioButton-input"
      type="radio"
      value=""
    />
    <div
      class="SxRadioButton-marker css-1389et7-SxRadioButton-marker"
    >
      <div
        class="SxRadioButton-markerDot css-nhzvrs-SxRadioButton-markerDot"
      />
    </div>
    <div
      class="SxRadioButton-content css-1rk6f29-SxRadioButton-content"
    >
      RadioButton
    </div>
  </label>
</DocumentFragment>
`;

exports[`RadioButton snapshots disabled 1`] = `
<DocumentFragment>
  <label
    class="SxRadioButton SxRadioButton_size_medium SxRadioButton_disabled css-w2gy44-SxRadioButton-root"
  >
    <input
      class="SxRadioButton-input css-1ykv6uc-SxRadioButton-input"
      disabled=""
      type="radio"
      value=""
    />
    <div
      class="SxRadioButton-marker css-1oszh0m-SxRadioButton-marker"
    >
      <div
        class="SxRadioButton-markerDot css-1kzafsd-SxRadioButton-markerDot"
      />
    </div>
    <div
      class="SxRadioButton-content css-1rk6f29-SxRadioButton-content"
    >
      RadioButton
    </div>
  </label>
</DocumentFragment>
`;

exports[`RadioButton snapshots hasError 1`] = `
<DocumentFragment>
  <label
    class="SxRadioButton SxRadioButton_size_medium SxRadioButton_hasError css-43x8eg-SxRadioButton-root"
  >
    <input
      class="SxRadioButton-input css-1ykv6uc-SxRadioButton-input"
      type="radio"
      value=""
    />
    <div
      class="SxRadioButton-marker css-4e3bbu-SxRadioButton-marker"
    >
      <div
        class="SxRadioButton-markerDot css-nhzvrs-SxRadioButton-markerDot"
      />
    </div>
    <div
      class="SxRadioButton-content css-1rk6f29-SxRadioButton-content"
    >
      RadioButton
    </div>
  </label>
</DocumentFragment>
`;

exports[`RadioButton snapshots large 1`] = `
<DocumentFragment>
  <label
    class="SxRadioButton SxRadioButton_size_large css-43x8eg-SxRadioButton-root"
  >
    <input
      class="SxRadioButton-input css-4aocnp-SxRadioButton-input"
      type="radio"
      value=""
    />
    <div
      class="SxRadioButton-marker css-6fp4t9-SxRadioButton-marker"
    >
      <div
        class="SxRadioButton-markerDot css-nhzvrs-SxRadioButton-markerDot"
      />
    </div>
    <div
      class="SxRadioButton-content css-k6soy6-SxRadioButton-content"
    >
      RadioButton
    </div>
  </label>
</DocumentFragment>
`;

exports[`RadioButton snapshots medium 1`] = `
<DocumentFragment>
  <label
    class="SxRadioButton SxRadioButton_size_medium css-43x8eg-SxRadioButton-root"
  >
    <input
      class="SxRadioButton-input css-4aocnp-SxRadioButton-input"
      type="radio"
      value=""
    />
    <div
      class="SxRadioButton-marker css-1389et7-SxRadioButton-marker"
    >
      <div
        class="SxRadioButton-markerDot css-nhzvrs-SxRadioButton-markerDot"
      />
    </div>
    <div
      class="SxRadioButton-content css-1rk6f29-SxRadioButton-content"
    >
      RadioButton
    </div>
  </label>
</DocumentFragment>
`;

exports[`RadioButton snapshots small 1`] = `
<DocumentFragment>
  <label
    class="SxRadioButton SxRadioButton_size_small css-43x8eg-SxRadioButton-root"
  >
    <input
      class="SxRadioButton-input css-4aocnp-SxRadioButton-input"
      type="radio"
      value=""
    />
    <div
      class="SxRadioButton-marker css-yddzja-SxRadioButton-marker"
    >
      <div
        class="SxRadioButton-markerDot css-uvghps-SxRadioButton-markerDot"
      />
    </div>
    <div
      class="SxRadioButton-content css-3chyuq-SxRadioButton-content"
    >
      RadioButton
    </div>
  </label>
</DocumentFragment>
`;
